mlflow/johnsnowlabs.py

```diff
@@ -0,0 +1,781 @@
+from mlflow.store.artifact.databricks_artifact_repo import \
+    DatabricksArtifactRepository
+from mlflow.tracking._model_registry import DEFAULT_AWAIT_MAX_SLEEP_SECONDS
+from mlflow.tracking.artifact_utils import (_download_artifact_from_uri,
+                                            _get_root_uri_and_artifact_path)
+from mlflow.utils import databricks_utils
+from mlflow.utils.docstring_utils import LOG_MODEL_PARAM_DOCS, format_docstring
+from mlflow.utils.environment import (_CONDA_ENV_FILE_NAME,
+                                      _CONSTRAINTS_FILE_NAME,
+                                      _PYTHON_ENV_FILE_NAME,
+                                      _REQUIREMENTS_FILE_NAME,
+                                      _mlflow_conda_env, _process_conda_env,
+                                      _process_pip_requirements, _PythonEnv)
+from mlflow.utils.file_utils import (TempDir,
+                                     shutil_copytree_without_file_permissions,
+                                     write_to)
+from mlflow.utils.model_utils import (_add_code_from_conf_to_system_path,
+                                      _get_flavor_configuration_from_uri,
+                                      _validate_and_copy_code_paths)
+from mlflow.utils.uri import (append_to_uri_path, dbfs_hdfs_uri_to_fuse_path,
+                              generate_tmp_dfs_path,
+                              get_databricks_profile_uri_from_artifact_uri,
+                              is_local_uri, is_valid_dbfs_uri)
+
+FLAVOR_NAME = "johnsnowlabs"
+
+_JOHNSNOWLABS_ENV_VARS = ('SECRET', 'AWS_ACCESS_KEY_ID', 'AWS_SECRET_ACCESS_KEY', 'SPARK_NLP_LICENSE')
+_JOHNSNOWLABS_MODEL_PATH_SUB = "jsl-model"
+_MLFLOWDBFS_SCHEME = "mlflowdbfs"
+_logger = logging.getLogger(__name__)
+
+
+def _validate_env_vars():
+    missing_env_vars = [os.environ.get(var) for var in _JOHNSNOWLABS_ENV_VARS]
+    if not missing_env_vars:
+        raise Exception(
+            f'Please set {",".join(missing_env_vars)}'
+            f' environment variables to your secret key before saving or logging your model')
+
+
+def _download_url(url, save_path):
+    file_name = url.split("/")[-1]
+    save_path = Path(save_path) / file_name
+    # Download the file from `url` and save it locally under `file_name`:
+    with urlopen(url) as response, open(str(save_path), "wb") as out_file:
+        shutil.copyfileobj(response, out_file)
+
+
+def get_default_pip_requirements():
+    """
+    :return: A list of default pip requirements for MLflow Models produced by this flavor.
+             Calls to :func:`save_model()` and :func:`log_model()` produce a pip environment
+             that, at minimum, contains these requirements.
+    """
+    from johnsnowlabs import settings
+    import pyspark
+    _SPARK_NLP_JSL_WHEEL_URI = "https://pypi.johnsnowlabs.com/{secret}/spark-nlp-jsl/spark_nlp_jsl-" \
+                               + f"{settings.raw_version_medical}-py3-none-any.whl"
```

Comment:
Storing a secret in a visible form like this (this URI is stored in `requirements.txt` and MLflow UI renders it) might not be the best idea.

Reply:
@C-K-Loan is there any mechanism by which the .whl file can be acquired and authentication can be provided via an environment config?